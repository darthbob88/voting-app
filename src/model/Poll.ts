export type Poll = {
    ID: string,
    name: string,
    creator: string,
    creation: number,
    expiration: number,
    method: string,
    candidates: string[],
    ballots: Ballot[]
}

export type Ballot = {
    ballotID: string,
    voterID: string,
    pollID: string,
    votes: any,
}