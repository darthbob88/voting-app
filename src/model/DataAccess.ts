import { Ballot, Poll } from "./Poll";

/*
 * Using local storage, just so I don't have to set up
 * Firebase until I'm more confident in the data structures. 
 */
//TODO: Set up Firebase, or whatever other data storage I decide to use.

export const createPoll = (newPoll: Poll) => {
    if (localStorage.getItem(newPoll.ID) != null) {
        throw new Error(`Invalid poll ${newPoll.ID}`);
    }
    localStorage.setItem(newPoll.ID, JSON.stringify(newPoll));
    return newPoll.ID;
}

export const retrievePoll = (pollID: string): Poll | null => {
    const jsonBlob = localStorage.getItem(pollID);
    if (jsonBlob == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }

    return JSON.parse(jsonBlob);

}

export const retrievePollList = (): string[] => {
    return Object.keys(localStorage);
}

export const castBallot = (newBallot: Ballot, pollID: string) => {
    const poll = retrievePoll(pollID);
    if (poll == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }
    poll.ballots.push(newBallot);
    localStorage.setItem(poll.ID, JSON.stringify(poll));
}