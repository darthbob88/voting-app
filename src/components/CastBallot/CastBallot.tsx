import { FormEvent, useState } from 'react';
import { castBallot, retrievePoll } from '../../model/DataAccess';
import { Ballot } from '../../model/Poll';

import styles from "./CastBallot.module.css";

type BallotProps = { pollID: string };

function CastBallot({ pollID }: BallotProps) {

    // But how do I type this for, say, IRV, where it needs a list of how each
    // candidate is ranked?
    // TODO: Work out how to type this.
    const [selectedOption, setSelectedOption] = useState<string>();

    const poll = retrievePoll(pollID);
    if (null == poll) {
        return <div>Invalid poll ID {pollID}</div>;
    }


    const renderBallot = (options: string[], method: string) => {
        if (method === "FPTP") {
            return <div className={styles.ballot}>
                {options.map(option => <label key={option}> <input

                    type="radio"
                    name="ballot"
                    value={option}
                    onChange={() => setSelectedOption(option)}
                    checked={selectedOption === option} />{option}</label>)}
            </div>
        } else if (method === "IRV") {
            return <div className={styles.ballot}> A series of dropdown menus to select the first/second/third/etc candidates</div>
        } else {
            return null;
        }
    }

    // TODO: This should redirect the user back to the home/poll page, and give some indication that we've cast a ballot.
    const submitBallot = (event: FormEvent) => {
        event.preventDefault()
        // TODO: Fix this to properly create a ballotID, voterID, etc.
        let newBallot: Ballot = { ballotID: "butts", voterID: "butts", votes: selectedOption, timestamp: Date.now(), pollID: pollID };
        castBallot(newBallot, pollID);
    }

    return (
        <form onSubmit={(event) => submitBallot(event)}><h3>Ballot for {poll.name}</h3>
            {renderBallot(poll.candidates, poll.method)}

            <div>Current value of ballot: {JSON.stringify(selectedOption)}</div>
            <div>Poll method is {poll.method}, options are {JSON.stringify(poll.candidates)}</div>

            {/* TODO: Disable the button until the ballot is valid. */}
            <button onClick={(event) => submitBallot(event)}>Cast Ballot</button>
        </form>
    );
}

export default CastBallot;