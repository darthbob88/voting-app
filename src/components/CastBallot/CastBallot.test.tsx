import { render } from '@testing-library/react';
import { createPoll } from '../../model/DataAccess';
import { testPoll } from '../../model/Poll';
import CastBallot from './CastBallot';

// TODO: Again, clear this once I have a proper database.
beforeEach(() => {
    localStorage.clear();
    createPoll(testPoll);

})

test('renders Ballot component', () => {
    const { getByText, container } = render(<CastBallot pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});

xtest('user can cast ballot with proper result', () => {
    const { getByText, container } = render(<CastBallot pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});

xtest('user cannot submit empty ballot', () => {
    const { getByText, container } = render(<CastBallot pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});



