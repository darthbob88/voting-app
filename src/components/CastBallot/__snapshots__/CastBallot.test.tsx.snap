// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Ballot component 1`] = `
<div>
  <form>
    <h3>
      Ballot for 
      pollName
    </h3>
    <div
      class="ballot"
    >
      <label>
         
        <input
          name="ballot"
          type="radio"
          value="apple"
        />
        apple
      </label>
      <label>
         
        <input
          name="ballot"
          type="radio"
          value="orange"
        />
        orange
      </label>
      <label>
         
        <input
          name="ballot"
          type="radio"
          value="banana"
        />
        banana
      </label>
    </div>
    <div>
      Current value of ballot: 
    </div>
    <div>
      Poll method is 
      FPTP
      , options are 
      ["apple","orange","banana"]
    </div>
    <button>
      Cast Ballot
    </button>
  </form>
</div>
`;
